rest/interceptors/src/main/java/com/connexta/spring/interceptor/VersionInterceptor.java

```diff
@@ -0,0 +1,290 @@
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.spring.interceptor;
+
+import com.connexta.discovery.rest.models.ErrorMessage;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.security.CodeSource;
+import java.security.ProtectionDomain;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nullable;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.input.ProxyInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Component;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+
+/**
+ * Spring interceptor capable of validating the version provided by the client with the version
+ * implemented by the server and return an error message accordingly. It will also automatically add
+ * the <code>Content-Version</code> header with the server version for the API called. This only
+ * happens in cases where the controller was called and generated a response and no other
+ * interceptors failed.
+ *
+ * <p>The server version for the API is retrieved from the dependency.proeprties file that must be
+ * included in the generated spring API using Apache's ServiceMix depends maven plugin.
+ */
+@Component
+public class VersionInterceptor implements HandlerInterceptor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(VersionInterceptor.class);
+
+  private static final String ACCEPT_VERSION = "Accept-Version";
+  private static final String CONTENT_VERSION = "Content-Version";
+
+  public static final String DEPENDENCIES_FILE = "META-INF/maven/dependencies.properties";
+  public static final String VERSION = "version";
+
+  private static final Map<Method, String> versions = new ConcurrentHashMap<>();
+
+  @Override
+  public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+      throws IOException {
+    final HandlerMethod handlerMethod = (HandlerMethod) handler;
```

Comment:
❓ Is this always a safe cast?

Reply:
Good question, everywhere I have seen interceptors built, they all do it this way. You tell me?